### PR TITLE
*:  collect coprocessor runtime stats for `explain analyze`

### PR DESCRIPTION
**Pull request by qw4990**
### What problem does this PR solve? 
SQL `explain analyze` can get execution information from TiKV(#8881).

### What is changed and how it works?
1. Update execution protocol (tipb) with TiKV to get its execution information.
2. Add `CopRuntimeStats` into `RuntimeStatsColl` to help to collect cop task execution information.
3. Modify `copIterator` and `RegionRequestSender` to set RPC callee address into `ExecDetails`.

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change

Side effects

 - Increased code complexity

Related changes

### CODE REVIEW

File: distsql/distsql_test.go [hunk 1]
```diff
@@ -85,6 +85,72 @@ func (s *testSuite) TestSelectNormal(c *C) {
...
+	c.Assert(err, IsNil)
+	result, ok := response.(*selectResult)
+	c.Assert(ok, IsTrue)
+	c.Assert(result.label, Equals, "dag")
+	c.Assert(result.sqlType, Equals, "general")
+	c.Assert(result.rowLen, Equals, len(colTypes))
+
+	// Test if planIDs is set into selectResult and selectResult can execute rightly.
+	selectResult, ok := response.(*selectResult)
+	if !ok {
```
Comment (lamxTyler) [inline distsql/distsql_test.go]: `c.Assert(ok, IsTrue)` is fine?

File: util/execdetails/execdetails.go [hunk 2]
```diff
@@ -141,6 +157,42 @@ func (e *RuntimeStatsColl) Get(planID string) *RuntimeStats {
...
+	}
+	var maxProcesses, totalRows, totalIters uint64
+	for _, stat := range stats {
+		if stat.timeProcessedNs > maxProcesses {
+			maxProcesses = stat.timeProcessedNs
+		}
+		totalRows += stat.numProducedRows
+		totalIters += stat.numIterations
+	}
+	return fmt.Sprintf("max proc ns:%v, total rows:%v, total iters:%v", maxProcesses, totalRows, totalIters)
```
Comment (lamxTyler) [inline util/execdetails/execdetails.go]: Can you explain more why we use the max execution times in a tikv? Sometimes, several cop tasks can be executed parallelly in one tikv.

File: distsql/distsql_test.go [hunk 3]
```diff
@@ -85,6 +85,72 @@ func (s *testSuite) TestSelectNormal(c *C) {
...
+		c.Fatal("not *selectResult")
+	}
+	if len(selectResult.copPlanIDs) != len(planIDs) {
+		c.Fatal("invalid copPlanIDs")
```
Comment (winoros) [inline distsql/distsql_test.go]: c.Assert(a, Equals, b)

File: distsql/distsql_test.go [hunk 4]
```diff
@@ -85,6 +85,72 @@ func (s *testSuite) TestSelectNormal(c *C) {
...
+	}
+	for i := range planIDs {
+		if selectResult.copPlanIDs[i] != planIDs[i] {
+			c.Fatal("invalid copPlanIDs")
```
Comment (winoros) [inline distsql/distsql_test.go]: ditto

File: executor/distsql.go [hunk 5]
```diff
@@ -302,7 +302,7 @@ func (e *IndexReaderExecutor) open(ctx context.Context, kvRanges []kv.KeyRange)
 		e.feedback.Invalidate()
 		return errors.Trace(err)
 	}
-	e.result, err = distsql.Select(ctx, e.ctx, kvReq, e.retTypes(), e.feedback)
+	e.result, err = distsql.SelectWithRuntimeStats(ctx, e.ctx, kvReq, e.retTypes(), e.feedback, getPhysicalPlanIDs(e.plans))
```
Comment (zz-jason) [inline executor/distsql.go]: We only need to collect runtime statistics when we are in `explain analyze` statement. For other normal SQL queries, we should directly call `distsql.Select()`?

File: util/execdetails/execdetails.go (hunk 2 above)
Comment (lamxTyler) [reply util/execdetails/execdetails.go]: What I concern is that it is the sum of execution time in a tikv, and two cop tasks may execute parallelly on a tikv, so the time may miss lead us.

Comment (qw4990) [reply util/execdetails/execdetails.go]: I just get what you mean.. It's really a problem.. I will fix it.

**Comment by qw4990** [comment]
@zz-jason @winoros @lamxTyler 
All comments have been addressed, PTAL~

File: util/execdetails/execdetails.go [hunk 6]
```diff
@@ -141,6 +161,46 @@ func (e *RuntimeStatsColl) Get(planID string) *RuntimeStats {
...
+	e.mu.Lock()
+	defer e.mu.Unlock()
+	stats, ok := e.copStats[planID]
+	if !ok {
+		return ""
+	}
+	var totalRows, totalIters, totalTasks uint64
+	procTimes := make([]uint64, 0, 32)
+	for _, instanceStats := range stats {
+		for _, stat := range instanceStats {
```
Comment (lamxTyler) [inline util/execdetails/execdetails.go]: Seems there is no need to differentiate the instance?

Comment (qw4990) [reply util/execdetails/execdetails.go]: It's not necessary now, but for further use, distinguishing the instance is useful.
For example, to trace which TiKV instance is the bottleneck of a query, we may list all tasks for each instance.

File: util/execdetails/execdetails.go [hunk 7]
```diff
@@ -141,6 +161,46 @@ func (e *RuntimeStatsColl) Get(planID string) *RuntimeStats {
...
+			procTimes = append(procTimes, stat.timeProcessedNs)
+			totalRows += stat.numProducedRows
+			totalIters += stat.numIterations
+			totalTasks++
+		}
+	}
+
+	n := len(procTimes)
+	sort.Slice(procTimes, func(i, j int) bool { return procTimes[i] < procTimes[j] })
+	return fmt.Sprintf("procNs max:%v, min:%v, p80:%v, p95:%v, rows:%v, iters:%v, tasks:%v",
```
Comment (lamxTyler) [inline util/execdetails/execdetails.go]: Can we also add the unit for time? It is hard to interpret something like `procNs max:129931, min:129931, p80:129931, p95:129931, rows:2, iters:2, tasks:1`.

File: executor/table_reader.go [hunk 8]
```diff
@@ -17,7 +17,7 @@ import (
 	"context"
 	"time"
 
-	opentracing "github.com/opentracing/opentracing-go"
+	"github.com/opentracing/opentracing-go"
```
Comment (winoros) [inline executor/table_reader.go]: revert this change.

File: executor/table_reader.go [hunk 9]
```diff
@@ -26,7 +26,7 @@ import (
 	"github.com/pingcap/tidb/table"
 	"github.com/pingcap/tidb/util/chunk"
 	"github.com/pingcap/tidb/util/ranger"
-	tipb "github.com/pingcap/tipb/go-tipb"
+	"github.com/pingcap/tipb/go-tipb"
```
Comment (winoros) [inline executor/table_reader.go]: ditto

File: store/mockstore/mocktikv/executor.go [hunk 10]
```diff
@@ -30,7 +31,7 @@ import (
 	"github.com/pingcap/tidb/types"
 	"github.com/pingcap/tidb/util/chunk"
 	"github.com/pingcap/tidb/util/codec"
-	tipb "github.com/pingcap/tipb/go-tipb"
+	"github.com/pingcap/tipb/go-tipb"
```
Comment (winoros) [inline store/mockstore/mocktikv/executor.go]: ditto

File: store/mockstore/mocktikv/cop_handler_dag.go [hunk 11]
```diff
@@ -37,7 +37,7 @@ import (
 	"github.com/pingcap/tidb/util/codec"
 	mockpkg "github.com/pingcap/tidb/util/mock"
 	"github.com/pingcap/tidb/util/timeutil"
-	tipb "github.com/pingcap/tipb/go-tipb"
+	"github.com/pingcap/tipb/go-tipb"
```
Comment (winoros) [inline store/mockstore/mocktikv/cop_handler_dag.go]: ditto

File: util/execdetails/execdetails.go [hunk 12]
```diff
@@ -141,6 +161,46 @@ func (e *RuntimeStatsColl) Get(planID string) *RuntimeStats {
...
+	}
+
+	n := len(procTimes)
+	sort.Slice(procTimes, func(i, j int) bool { return procTimes[i] < procTimes[j] })
+	return fmt.Sprintf("proc max:%v, min:%v, p80:%v, p95:%v, rows:%v, iters:%v, tasks:%v",
+		procTimes[n-1], procTimes[0], procTimes[n*4/5], procTimes[n*19/20], totalRows, totalIters, totalTasks)
+}
+
 // Exists checks if the planID exists in the stats collection.
 func (e *RuntimeStatsColl) Exists(planID string) bool {
```
Comment (XuHuaiyu) [inline util/execdetails/execdetails.go]: s/ Exists/ ExistsRootStats
s/ Get/ GetRootStats

File: distsql/select_result.go [hunk 13]
```diff
@@ -68,6 +68,10 @@ type selectResult struct {
 	feedback     *statistics.QueryFeedback
 	partialCount int64 // number of partial results.
 	sqlType      string
+
+	// copPlanIDs contains all copTasks' planIDs,
+	// which help to collect copTasks' runtime stats.
+	copPlanIDs []string
```
Comment (XuHuaiyu) [inline distsql/select_result.go]: We can remove this and change the definition of `copStats map[string]map[string][]*CopRuntimeStats` to `copStats []map[string][]*CopRuntimeStats` since we can use the ordinal to find the corresponding element in copStats.

File: util/execdetails/execdetails.go [hunk 14]
```diff
@@ -141,6 +161,46 @@ func (e *RuntimeStatsColl) Get(planID string) *RuntimeStats {
 	return runtimeStats
 }
 
+// RecordOneCopTask records a specific cop tasks's execution detail.
+func (e *RuntimeStatsColl) RecordOneCopTask(planID, address string,
+	timeProcessedNs, numProducedRows, numIterations uint64) {
+	e.mu.Lock()
```
Comment (XuHuaiyu) [inline util/execdetails/execdetails.go]: It seems that lock is not needed for copRuntimeStats.

File: util/execdetails/execdetails.go [hunk 15]
```diff
@@ -141,6 +161,46 @@ func (e *RuntimeStatsColl) Get(planID string) *RuntimeStats {
...
+		e.copStats[planID] = make(map[string][]*CopRuntimeStats, 8)
+	}
+
+	e.copStats[planID][address] = append(e.copStats[planID][address],
+		&CopRuntimeStats{timeProcessedNs, numProducedRows, numIterations})
+}
+
+// CopSummary gets a summary of the cop task's execution information.
+func (e *RuntimeStatsColl) CopSummary(planID string) string {
+	e.mu.Lock()
```
Comment (XuHuaiyu) [inline util/execdetails/execdetails.go]: ditto

File: distsql/select_result.go [hunk 16]
```diff
@@ -167,6 +172,22 @@ func (r *selectResult) getSelectResp() error {
...
 }
 
+func (r *selectResult) updateCopRuntimeStats(callee string) {
+	if r.ctx.GetSessionVars().StmtCtx.RuntimeStatsColl == nil ||
+		callee == "" || len(r.selectResp.GetExecutionSummaries()) != len(r.copPlanIDs) {
+		return
+	}
+
+	for i, detail := range r.selectResp.GetExecutionSummaries() {
+		if detail != nil && detail.TimeProcessedNs != nil &&
```
Comment (XuHuaiyu) [inline distsql/select_result.go]: when will TimeProcessedNs or NumProducedRows or NumIterations be nil?

File: planner/core/common_plans.go [hunk 17]
```diff
@@ -530,10 +530,15 @@ func (e *Explain) prepareOperatorInfo(p PhysicalPlan, taskType string, indent st
 	row := []string{e.prettyIdentifier(p.ExplainID(), indent, isLastChild), count, taskType, operatorInfo}
 	if e.Analyze {
 		runtimeStatsColl := e.ctx.GetSessionVars().StmtCtx.RuntimeStatsColl
-		if runtimeStatsColl.Exists(p.ExplainID()) {
+		copTaskExecDetail := runtimeStatsColl.CopSummary(p.ExplainID())
+		// There maybe some mock information for cop task to let runtimeStatsColl.Exists(p.ExplainID()) is true.
```
Comment (XuHuaiyu) [inline planner/core/common_plans.go]: May we remove all the mock info when tikv supports runtimeStats?

File: util/execdetails/execdetails.go [hunk 18]
```diff
@@ -141,6 +161,46 @@ func (e *RuntimeStatsColl) Get(planID string) *RuntimeStats {
 	return runtimeStats
 }
 
+// RecordOneCopTask records a specific cop tasks's execution detail.
+func (e *RuntimeStatsColl) RecordOneCopTask(planID, address string,
```
Comment (winkyao) [inline util/execdetails/execdetails.go]: Why you do not just use tipb.ExecutorExecutionSummary as the parameter?

File: util/execdetails/execdetails.go [hunk 19]
```diff
@@ -124,9 +133,20 @@ type RuntimeStats struct {
 	rows int64
 }
 
+// CopRuntimeStats collects info of one executor executed on TiKV.
```
Comment (XuHuaiyu) [inline util/execdetails/execdetails.go]: Why don't we reuse RuntimeStats?

File: distsql/select_result.go (hunk 16 above)
Comment (winkyao) [reply distsql/select_result.go]: Just a protect?

File: distsql/select_result.go [hunk 20]
```diff
@@ -167,6 +172,22 @@ func (r *selectResult) getSelectResp() error {
 	}
 }
 
+func (r *selectResult) updateCopRuntimeStats(callee string) {
+	if r.ctx.GetSessionVars().StmtCtx.RuntimeStatsColl == nil ||
+		callee == "" || len(r.selectResp.GetExecutionSummaries()) != len(r.copPlanIDs) {
```
Comment (zz-jason) [inline distsql/select_result.go]: if `len(r.selectResp.GetExecutionSummaries()) != len(r.copPlanIDs)`, we should log an error

File: distsql/select_result.go (hunk 13 above)
Comment (zz-jason) [reply distsql/select_result.go]: Yeah, if this is removed from `selectResult`, we can remove `SelectWithRuntimeStats()` as well.

File: distsql/select_result.go (hunk 16 above)
Comment (qw4990) [reply distsql/select_result.go]: Yes, just a protection to prevent from panic caused by wrong result from TiKV.

File: distsql/select_result.go (hunk 20 above)
Comment (qw4990) [reply distsql/select_result.go]: fixed.

File: util/execdetails/execdetails.go (hunk 19 above)
Comment (qw4990) [reply util/execdetails/execdetails.go]: Good suggestion.
I have updated it and `RuntimeStats` is reused now.

File: util/execdetails/execdetails.go (hunk 12 above)
Comment (qw4990) [reply util/execdetails/execdetails.go]: fixed

File: distsql/distsql.go [hunk 21]
```diff
@@ -73,6 +73,21 @@ func Select(ctx context.Context, sctx sessionctx.Context, kvReq *kv.Request, fie
 	}, nil
 }
 
+// SelectWithRuntimeStats sends a DAG request, returns SelectResult.
+// The different from Select is SelectWithRuntimeStats will set copPlanIDs into selectResult,
```
[the remark this answers is not in the document]
Comment (qw4990) [reply distsql/distsql.go]: fixed

File: planner/core/common_plans.go (hunk 17 above)
Comment (qw4990) [reply planner/core/common_plans.go]: It's removed now.

File: util/execdetails/execdetails.go (hunk 18 above)
Comment (qw4990) [reply util/execdetails/execdetails.go]: Good suggestion. Its parameters have been updated now.

File: util/execdetails/execdetails.go (hunk 15 above)
Comment (qw4990) [reply util/execdetails/execdetails.go]: This lock is removed.

File: distsql/select_result.go [hunk 22]
```diff
@@ -167,6 +173,27 @@ func (r *selectResult) getSelectResp() error {
...
+		return
+	}
+
+	for i, detail := range r.selectResp.GetExecutionSummaries() {
+		if detail != nil && detail.TimeProcessedNs != nil &&
+			detail.NumProducedRows != nil && detail.NumIterations != nil {
+			planID := r.copPlanIDs[i]
+			r.ctx.GetSessionVars().StmtCtx.RuntimeStatsColl.
+				RecordOneCopTask(planID, callee, detail)
+			//, int64(*detail.TimeProcessedNs), int64(*detail.NumProducedRows), int32(*detail.NumIterations))
```
Comment (XuHuaiyu) [inline distsql/select_result.go]: useless line?

File: distsql/select_result.go [hunk 23]
```diff
@@ -167,6 +173,27 @@ func (r *selectResult) getSelectResp() error {
...
+	}
+	if len(r.selectResp.GetExecutionSummaries()) != len(r.copPlanIDs) {
+		log.Errorf("invalid cop task execution summaries length, expected: %v, received: %v",
+			len(r.copPlanIDs), len(r.selectResp.GetExecutionSummaries()))
+		return
+	}
+
+	for i, detail := range r.selectResp.GetExecutionSummaries() {
+		if detail != nil && detail.TimeProcessedNs != nil &&
+			detail.NumProducedRows != nil && detail.NumIterations != nil {
```
Comment (XuHuaiyu) [inline distsql/select_result.go]: What if the 3 attributes are nil? It seems that some info will be lost, I think this is not acceptable?

File: planner/core/common_plans.go (hunk 17 above)
Comment (XuHuaiyu) [reply planner/core/common_plans.go]: 1. Can we refactor this `if-else` to :
if runtimeStatsColl.ExistsRootStats(p.ExplainID()){
   row = append(row, runtimeStatsColl.GetRootStats(p.ExplainID()).String())
}else {
   row = append(row, runtimeStatsColl.CopSummary(p.ExplainID()))
}

2. Can we implement a `String()` function instead of `RuntimeStatsColl.CopSummary` for CopRuntimeStats? If so, we can use `row = append(row, runtimeStatsColl.GetCopStats(p.ExplainID()).String())`

File: distsql/select_result.go (hunk 23 above)
Comment (qw4990) [reply distsql/select_result.go]: emm.. Is there anything we can do more? The only thing I can think of is printing some logs.. Do you have any suggestion?

**Comment by qw4990** [comment]
All comments have been addressed, PTAL~ @XuHuaiyu

**Comment by XuHuaiyu** [comment]
The info of TableScan_17 is still confused.
``` sql
CREATE TABLE `t` (
  `i` int(11) DEFAULT NULL,
  `a` int(11) DEFAULT NULL,
  KEY `i` (`i`)
);
insert into t values(1, null), (2, null), (3, null);
insert into t values(1, null), (2, null), (3, null);
insert into t values(1, null), (2, null), (3, null);

tidb> explain analyze select * from t where i < 5 order by 1;
+----------------------+---------+------+-----------------------------------------------------------------+-------------------------------------------------------------------------------------------------+
| id                   | count   | task | operator info                                                   | execution info
                                   |
+----------------------+---------+------+-----------------------------------------------------------------+-------------------------------------------------------------------------------------------------+
| Projection_19        | 3323.33 | root | test.t.i, test.t.a                                              | time:513.623µs, loops:2, rows:9 |
| └─IndexLookUp_18     | 3323.33 | root |                                                                 | time:463.261µs, loops:2, rows:9 |
|   ├─IndexScan_16     | 3323.33 | cop  | table:t, index:i, range:[-inf,5), keep order:true, stats:pseudo | proc max:112.144µs, min:112.144µs, p80:112.144µs, p95:112.144µs, rows:18, iters:10, tasks:1     |
|   └─TableScan_17     | 3323.33 | cop  | table:t, keep order:false, stats:pseudo                         | time:0s, loops:0, rows:9 -- This line is confused.  |
+----------------------+---------+------+-----------------------------------------------------------------+-------------------------------------------------------------------------------------------------+
4 rows in set (0.00 sec)
```

File: util/execdetails/execdetails.go [hunk 24]
```diff
@@ -108,10 +111,54 @@ func (d ExecDetails) String() string {
...
 
+// CopRuntimeStats collects cop tasks' execution info.
+type CopRuntimeStats struct {
+	sync.Mutex
+
+	// Usually, CopTasks are executed on TiKV cluster consist of multiple instances,
+	// so coprocessors' addresses must be taken into account.
+	// And Sometimes, several cop tasks of one operator can be executed in a same TiKV instance,
+	// which depend on how we split cop tasks and the distribution of data regions.
+	// So we have to use a list to maintain all tasks executed on each instance.
```
[the remark this answers is not in the document]
Comment (qw4990) [reply util/execdetails/execdetails.go]: These comments have been updated.

File: util/execdetails/execdetails.go [hunk 25]
```diff
@@ -108,10 +111,54 @@ func (d ExecDetails) String() string {
...
+			procTimes = append(procTimes, time.Duration(stat.consume)*time.Nanosecond)
+			totalRows += stat.rows
+			totalIters += stat.loop
+			totalTasks++
+		}
+	}
+
+	n := len(procTimes)
+	sort.Slice(procTimes, func(i, j int) bool { return procTimes[i] < procTimes[j] })
+	return fmt.Sprintf("proc max:%v, min:%v, p80:%v, p95:%v, rows:%v, iters:%v, tasks:%v",
```
[the remark this answers is not in the document]
Comment (qw4990) [reply util/execdetails/execdetails.go]: Good suggestion, I will do this in another PR.

File: util/execdetails/execdetails.go [hunk 26]
```diff
@@ -108,10 +111,54 @@ func (d ExecDetails) String() string {
...
+		procTimes[n-1], procTimes[0], procTimes[n*4/5], procTimes[n*19/20], totalRows, totalIters, totalTasks)
```
[the remark this answers is not in the document]
Comment (qw4990) [reply util/execdetails/execdetails.go]: The output when there is one cop task has been updated, and now `process time`, `loops` and `rows` will be outputted like the output of root task.

**Comment by qw4990** [comment]
The problem you mentioned above has been resolved now.
The reason is that we [forbid collecting runtime stats](https://github.com/pingcap/tidb/pull/9057/files#diff-b28d935951b1bf48cb553e8e4bb9dedbL509) on `IndexLookUp.TableWalker`, so the runtime stats of `TableScan_17` is wrong.
PTAL~ @XuHuaiyu

File: planner/core/common_plans.go [hunk 27]
```diff
@@ -529,10 +529,12 @@ func (e *Explain) prepareOperatorInfo(p PhysicalPlan, taskType string, indent st
 	row := []string{e.prettyIdentifier(p.ExplainID(), indent, isLastChild), count, taskType, operatorInfo}
 	if e.Analyze {
 		runtimeStatsColl := e.ctx.GetSessionVars().StmtCtx.RuntimeStatsColl
-		if runtimeStatsColl.Exists(p.ExplainID()) {
-			row = append(row, runtimeStatsColl.Get(p.ExplainID()).String())
-		} else {
-			row = append(row, "") //TODO: wait collect more executor info from tikv
+		// There maybe some mock information for cop task to let runtimeStatsColl.Exists(p.ExplainID()) is true.
```
Comment (XuHuaiyu) [inline planner/core/common_plans.go]: Will the case mentioned in this comment happen for now?

File: util/execdetails/execdetails.go [hunk 28]
```diff
@@ -15,6 +15,8 @@ package execdetails
 
 import (
 	"fmt"
+	"github.com/pingcap/tipb/go-tipb"
```
Comment (XuHuaiyu) [inline util/execdetails/execdetails.go]: format this import

File: util/execdetails/execdetails.go [hunk 29]
```diff
@@ -126,26 +177,53 @@ type RuntimeStats struct {
...
 		runtimeStats = &RuntimeStats{}
-		e.stats[planID] = runtimeStats
+		e.rootStats[planID] = runtimeStats
 	}
 	return runtimeStats
 }
 
-// Exists checks if the planID exists in the stats collection.
-func (e *RuntimeStatsColl) Exists(planID string) bool {
+// GetCopStats gets the CopRuntimeStats specified by planID
```
Comment (XuHuaiyu) [inline util/execdetails/execdetails.go]: add a `.` at the end

File: util/execdetails/execdetails_test.go [hunk 30]
```diff
@@ -14,6 +14,7 @@
 package execdetails
 
 import (
+	"github.com/pingcap/tipb/go-tipb"
```
Comment (XuHuaiyu) [inline util/execdetails/execdetails_test.go]: format this import.

File: planner/core/common_plans.go (hunk 27 above)
Comment (qw4990) [reply planner/core/common_plans.go]: Yes.
The root task stats are created at `newBaseExecutor`, and it is unknown weather the executor is built by a cop task or a root task, so the default action now is creating a root stats for it.

Maybe we can introduce one more parameter into `newBaseExecutor` to indicate the type of task, but I prefer to do it in another PR because this PR is big enough...

File: util/execdetails/execdetails.go (hunk 28 above)
Comment (qw4990) [reply util/execdetails/execdetails.go]: It's my fault. 
All import problems have been fixed.

File: util/execdetails/execdetails_test.go (hunk 30 above)
Comment (qw4990) [reply util/execdetails/execdetails_test.go]: fixed.

File: util/execdetails/execdetails.go (hunk 29 above)
Comment (qw4990) [reply util/execdetails/execdetails.go]: fixed.

File: planner/core/common_plans.go (hunk 27 above)
Comment (XuHuaiyu) [reply planner/core/common_plans.go]: Can this comment be removed after implementing analyze in TiKV?

Comment (qw4990) [reply planner/core/common_plans.go]: I'm afraid not.
For each cop task, an empty root stats will be created for it in `newBaseExecutor`.
So unless we fix it in `newBaseExecutor`, this problem will always exist.

Comment (XuHuaiyu) [reply planner/core/common_plans.go]: Do you mean rootStatsMap and copStatsMap will contain the same explainID?

Comment (qw4990) [reply planner/core/common_plans.go]: yes...

File: util/execdetails/execdetails.go (hunk 28 above)
Comment (XuHuaiyu) [inline util/execdetails/execdetails.go]: format this

**Comment by XuHuaiyu** [comment]
PTAL @zz-jason
